schism/page_info.c

```diff
@@ -729,10 +735,133 @@ static void info_draw_note_dots(int base, int height, int active, int first_chan
 	}
 }
 
+#ifdef ENABLE_WAVEFORMVIS
+static void _calculate_channel_layout(int num_channels, int width, int height, int *num_rows, int *num_columns)
+{
+	for (*num_rows = 1; *num_rows < 8; (*num_rows)++)
+	{
+		int channel_width, channel_height;
+		int aspect; // scaled up by 100
+
+		*num_columns = (num_channels + *num_rows - 1) / *num_rows;
+
+		channel_width = width / *num_columns;
+		channel_height = height / *num_rows;
+
+		aspect = channel_height * 100 / channel_width;
+
+		if (aspect < 180)
+			break;
+	}
+}
+
+static void info_draw_waveform_recalculate(struct info_window *window)
+{
+	window->overlay.x1 = 3;
+	window->overlay.x2 = 77;
+	window->overlay.y1 = window->first_row + 1;
+	window->overlay.y2 = window->first_row + window->height - 2;
+
+	vgamem_ovl_alloc(&window->overlay);
+}
+
+#define WAVEFORM_BOX_OUTLINE_COLOUR 1
+#define WAVEFORM_COLOUR 5
+
+static void info_draw_waveform(struct info_window *window, int base, int height, int active, int first_channel)
+{
+	int num_channels = window->type_def->channels;
+
+	if (num_channels == 0) {
+		// mix output
+		first_channel = MAX_VOICES + 1;
+		num_channels = current_song->flags & SONG_NOSTEREO ? 1 : 2;
+	}
+
+	int num_rows, num_columns;
+	int chan, chan_pos;
+
+	if ((window->overlay.width <= 0) || (window->overlay.height <= 0))
+		return;
+
+	_calculate_channel_layout(num_channels, window->overlay.width, window->overlay.height, &num_rows, &num_columns);
+
+	vgamem_ovl_clear(&window->overlay, 0);
+
+	for (chan = MAX(0, first_channel - 1), chan_pos = 0; chan_pos < num_channels; chan++, chan_pos++) {
+		int row = chan_pos / num_columns;
+		int col = chan_pos % num_columns;
+
+		int x1 = col * window->overlay.width / num_columns;
+		int y1 = row * window->overlay.height / num_rows;
+		int x2 = (col + 1) * window->overlay.width / num_columns - 1;
+		int y2 = (row + 1) * window->overlay.height / num_rows - 1;
+
+		uint8_t *recent_samples;
+		int idx;
+
+		for (int y = y1 | 1; y <= y2; y += 2)
+			vgamem_ovl_drawpixel(&window->overlay, x2, y, WAVEFORM_BOX_OUTLINE_COLOUR);
+		for (int x = x1 | 1; x <= x2; x += 2)
+			vgamem_ovl_drawpixel(&window->overlay, x, y2, WAVEFORM_BOX_OUTLINE_COLOUR);
+
+		if ((chan >= MAX_VOICES) || current_song->voices[chan].current_sample_data) {
+			recent_samples = RECENT_SAMPLE_BUFFER(chan);
```

Comment:
what's going on here? why not use the vgamem sample drawing functions?

Reply:
If anything, the code should just be changed to allow for any arbitrary coordinates (which means adding x,y,width,height params)